### PR TITLE
Fix Two Bugs in the Text Layer

### PR DESCRIPTION
**Pull request by speedplane**
## The Problem

The x position of some elements in the text layer are not being calculated properly.  For example, try selecting [the fourth line of this PDF](http://www.docdroid.net/file/view/tsea/usa-v-serrapio-page2.pdf). Selection will be difficult and may not work at all. If you load the same PDF into Adobe Reader, it works fine. The text layer divs are not being positioned properly. 

The problem is apparent in the image below, where all the divs are drawn with borders. On the fourth line, there are two divs, which _should_ be adjacent to each other, but are instead overlapping each other. Further, the second div on the fourth line does not extend to the end of the line. The fifth line has similar problems.

image
## The Solution

The code that causes this bug is relatively deep into the heart of the PDF processing.  However, it appears that  there are two bugs. First, we are scaling a variable twice instead of once. Second, the textState is not being calculated properly when there is whitespace.

This PR fixes those two issues. Now any PDF `TJ` command which has `wordSpacing` and actual spaced words within it will be properly spaced in the text layer.  You can [play with this fix here](https://www.docketalarm.com/cases/US_Court_of_Appeals_Eleventh_Circuit/12-14897/USA_v._Joaquin_Serrapio_Jr./Order-06-18-14/).
## Testing

I plan on putting together a regression test for this PR.  In particular, I want to build a test that ensures that no displayed text elements overlap each other on the screen. While it's true that some PDFs do have overlapping text, most do not, and a test that checks specific PDFs can help find some bugs. However, such a test will rely on changes in #5545, which adds a quadtree structure. Once (if!) that gets accepted, I will add a test here.

### CODE REVIEW

**Comment by speedplane** [comment]
Hmm... just found #5688, which appears to implement one of the two bug fixes described above. It also adds another bug fix. I'll incorporate.

File: src/core/evaluator.js [hunk 1]
```diff
@@ -971,7 +972,9 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         for (var i = 0; i < glyphs.length; i++) {
           var glyph = glyphs[i];
           if (!glyph) { // Previous glyph was a space.
-            width += textState.wordSpacing * textState.textHScale;
+            var sp_tx = textState.wordSpacing * textState.textHScale;
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: Nit: please remember to use more descriptive (and complete) variable names :-)
Also don't forget that variables should be lowerCamelCase, for reference: https://github.com/mozilla/pdf.js/wiki/Style-Guide#naming.

**Comment by Snuffleupagus** [comment]
For testing purposes, you might also want to look into adding `text` tests, see https://github.com/mozilla/pdf.js/blob/master/test/test_manifest.json.

**Comment by speedplane** [comment]
Made the style changes. Also, extended this PR to fix related issues with vertical text.

@Snuffleupagus I took a look at the `"type": "text", they don't appear to do all that much.  I can build upon them though. Once the quadTree structure is added (PR #5545), we can add a flag to run tests over all elements to ensure that elements do not overlap.

**Comment by speedplane** [comment]
@Snuffleupagus Actually... those tests do appear to do more than I expected. Given that the tests failed in #5688, and I make a similar change, I'm betting that this will cause them to fail as well.  Where do you update the reference files?

**Comment by timvandermeij** [comment]
@speedplane You can generate reference images yourself using the steps from https://github.com/mozilla/pdf.js/wiki/Contributing. On here we will update the reference images once we merge a pull request.

**Comment by timvandermeij** [comment]
Would you say this PR supersedes #5688 or can we review/merge that one first?

**Comment by speedplane** [comment]
Yes, this one fixes the bug in #5688, and includes one more.

**Comment by Hengjie** [comment]
Yeah, this PR fixes the same problem that I was fixing. We're doing it slightly different, but to the same effect. Feel free to merge this one since it incorporates more fixes :)

**Comment by timvandermeij** [comment]
Fixes #5687

File: src/core/evaluator.js [hunk 2]
```diff
@@ -1109,16 +1118,19 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
                 if (typeof items[j] === 'string') {
                   buildTextGeometry(items[j], textChunk);
                 } else {
+                  // See Section 5.3.2. Shift right/down by unit thousanths.
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Nit: unit `thousandths`

**Comment by timvandermeij** [comment]
@speedplane This patch looks good so far. Please fix the mentioned nit and add a `text` test case to avoid future regressions.

**Comment by timvandermeij** [comment]
@speedplane Do you have time to complete this patch?

**Comment by speedplane** [comment]
I want to. I probably won't have time for another few weeks though.

Michael Sander
michael.sander@gmail.com
607-227-9859

**Comment by nschloe** [comment]
Is this PR still being worked on?

**Comment by speedplane** [comment]
I've put in my calendar to start working on this again starting Sep. 7

**Comment by yurydelendik** [comment]
The broken PDF was fixed by #6588. Thanks.

**Pull request closed without merging**